lib/ASTGen/Sources/ASTGen/ASTGen.swift

```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+  func visit(_ node: GenericWhereClauseSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: EnumCaseParameterClauseSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: InheritedTypeListSyntax?) -> BridgedArrayRef {
+    guard let node else {
+      return .init()
+    }
+
+    return self.visit(node) 
+  }
+
+  @inline(__always)
+  func visit(_ node: PrecedenceGroupNameListSyntax?) -> BridgedArrayRef {
+    guard let node else {
+      return .init()
+    }
+
+    return self.visit(node)
+  }
+}
+
+extension SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Element) -> T) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      var currentAddress = baseAddress
+      for element in self {
+        currentAddress.initialize(to: transform(element))
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+
+  /// Like ``Sequence.compactMap(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func compactMap<T>(in astgen: ASTGenVisitor, _ transform: (Element) -> T?) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      var currentAddress = baseAddress
+      for element in self {
+        guard let transformed = transform(element) else {
+          continue
+        }
+
+        currentAddress.initialize(to: transformed)
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+}
+
+extension Optional where Wrapped: SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  @inline(__always)
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Wrapped.Element) -> T) -> BridgedArrayRef {
+    switch self {
+    case .none: return .init()
+    case .some(let unwrapped): return unwrapped.map(in: astgen, transform)
+    }
+  }
+}
+
+extension Collection {
+  /// Like ``Sequence.reversed()``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func reversed(in astgen: ASTGenVisitor) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(Element.self, count: self.count).baseAddress!
+
+    do {
+      var currentAddress = baseAddress
+      for element in self.lazy.reversed() {
+        currentAddress.initialize(to: element)
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+}
+
```

Comment:
> ```diff
> - let memberComponents = reverseMemberComponents.reversed(in: self)
> + let memberComponents = reverseMemberComponents.reversed().map(in: self) { $0 }
> ```

I really want to avoid intermediate arrays, mostly because of the performance smell but in a way I also see them as defeating the purpose of bump allocation—we would still be reference-counting and allocating just as much.

> As a bonus, we could define 
> 
> ```swift
> extension Collection {
>   func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
>     return self.map(in: astgen) { $0 }
>   } 
> }
> ```

I [had a very similar idea](https://github.com/apple/swift/pull/67111#discussion_r1316084733) that enforces lazy array bridging. This helps to prevent eager operations to a certain extent by guiding the programmer. The problem of doing this on `LazyCollectionProtocol` is we need a sensible upper bound for how much to allocate, but using `count` on lazy wrappers can be very costly for compact maps and such. Would you be up for this sort of hack:

1. Add a `ASTGenLazyCollection` protocol with a single `maxCount` requirement.
1. The default implementation of `maxCount` can just return `count`.
1. Conform the standard library lazy wrappers we use to `ASTGenLazyCollectionProtocol`
1. ```swift
   extension ASTGenLazyCollectionProtocol {
     func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
       if self.isEmpty {
         return .init()
       }
    
       let buffer = astgen.allocator.allocate(Element.self, count: self.maxCount)
       _ = buffer.initialize(from: self)
    
       return .init(data: buffer.baseAddress!, numElements: SwiftInt(self.maxCount))
     }
   }
   ```

This trades the performance boost from defining custom operations vs. using lazy wrappers for a consistent, low-maintenance, and more or less future-proof pattern.

Reply:
> I really want to avoid intermediate arrays, mostly because of the performance smell but in a way I also see them as defeating the purpose of bump allocation—we would still be reference-counting and allocating just as much.

Ah, damn. I forgot about the intermediate array creation while reviewing this. I’m fine with keeping it. Could you add a doc comment to `reversed(in:)` that highlights that the method exists for exactly this performance reason?

> Conform the standard library lazy wrappers we use to `ASTGenLazyCollectionProtocol`

Which types would you conform to `ASTGenLazyCollectionProtocol`? For example `array.lazy.reversed()` is a `ReversedCollection<LazySequence<UnderlyingSequence>>` and I don’t think we can guarantee efficient `maxCount` access on that.

Reply:
> Which types would you conform to `ASTGenLazyCollectionProtocol`? For example `array.lazy.reversed()` is a `ReversedCollection<LazySequence<UnderlyingSequence>>` and I don’t think we can guarantee efficient `maxCount` access on that.

Arrgh! `ReversedCollection._base` is public, which made me think the bases in other wrappers are too. I guess this nut is unlikely to ever crack nicely unless we integrate `maxCount` into the stdlib like so:

* `LazySequenceProtocol` gets a `base` requirement
* `Sequence` gets a `maxCount` requirement with default implementations of  `Int.max` on `Sequence`, `self.count` on `Collection`, and `base.maxCount` on `LazySequenceProtocol`.
* `ReversedCollection` conforms to `LazySequenceProtocol`

Since (as Rintaro pointed out below) `reversed` is already lazy on `BidirectionalCollection`, let’s do `Collection.bridgedArray(in:)` and make an exception for `compactMap`. As much as I dislike having to keep `compactMap(in:)`, I would rather not have us calling `count` on a lazy `compactMap`.

Reply:
Sounds good 👍🏽. I can’t tell you why but `compactMap` also triggered me a lot less than `reversed`.